### PR TITLE
Remove Or, And, Not + related fixes

### PR DESCRIPTION
**Pull request by olsaarik**
This PR includes cleanup of unused nodes and related fixes. The `SymbolicRegexNodeKind` elements `Or`, `And` and `Not` as well as any related classes/functions have been removed. This includes `SymbolicRegexSet`, `TransitionRegex` and `SymbolicNFA`.

`OrderedOr` has been renamed to `Alternate`, to match the name of the parse tree node. Their semantics are the same.

`SaveDGML`, which depended on `Or`, has been reworked by pulling out the exploration logic into `SymbolicRegexMatcher.Explore`, which does the exploration in the main transition structures of the matcher. This means that the exploration can also be used to pre-calculate all the derivatives instead of doing it on the fly, which might be useful for future optimizations (e.g. combining `Compiled` and `NonBacktracking`).

`GenerateRandomMembers` is now `SampleMatches` and has also been rewritten to use the transition logic from `SymbolicRegexMatcher`. Negative sampling has been removed due to not having support for `Not` anymore.

`SymbolicRegexMatcher` is now a partial class and `SaveDGML`, `Explore`, and `SampleMatches` are implemented in separate files as additions to the matcher.

`FixedLengthMarker` support had several problems that this PR fixes:

- Logic to extract the fixed length out of an end state only supported the now-removed Or. The main logic is now in `SymbolicRegexNode.ResolveFixedLength`. The new function supports finding the fixed length marker on the path that the backtracking matcher would accept a match and also supports conditional nullability. For example, patterns like `abc|$(4)|(3)` would correctly resolve to either 4 or 3 depending on the context.
- Logic to add fixed length markers in `RegexNodeConverter` didn't support the structures actually present now that we have capture start and end markers. The logic is now in a separate transformation function `SymbolicRegexNode.AddFixedLengthMarkers`, which made it much easier to handle the various cases. It does introduce some additional allocations, as parts of the tree may be rebuilt as markers are added.

### CODE REVIEW

**Comment by stephentoub** [comment]
This is very pretty:
image

File: src/libraries/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System.Text.RegularExpressions.csproj:69]: Nit: can you sort these?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs:32]: It's great that you were able to fix the fixed-length markers.  Do we know if this was contributing to some of the perf slowdowns that had been measured?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs:139]: > It also counts the number of transitions found.

What is this referring to?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs:148]: Do we need "Span" in the name of these helpers, or could they just be GetDfaDeltasFor and GetNfaDeltasFor?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs:166]: Nit: var => int

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Explore.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Explore.cs:55]:
```suggestion

        DfaLimitReached:
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Explore.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Explore.cs:72]:
```suggestion

                while (toExplore.Count > 0)
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Explore.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Explore.cs:82]: var => whatever this is

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs:939]: Does "BreakUpAlternation" mean "EnumerateAlternationBranches"?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1021]: It may not actually matter since we hopefully never hit this path, but I'm wondering if we should just make this return rather than forking to a new thread.  When the correctness of the algorithm depends on continuing, we need to fork like this, but this path is just an optimization, right?  In which case we could just stop optimizing if the expression is so complex that you end up being too deep?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1748]:
```suggestion
                foreach (char c in value.ToString(CultureInfo.InvariantCulture))
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1756]:
```suggestion
                foreach (char c in value.ToString(CultureInfo.InvariantCulture))
```

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:2201]: "so the left child is always an element of the alternation"... and is not itself an alternation, yes?

**Review by stephentoub — APPROVED** [review]
Thanks!

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Debug.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Debug.cs:62]: Yes that's correct. I'll add a clarifying remark on that.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs:139]: Ah, the comment's fallen out of sync during development. Thanks!

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Dgml.cs:148]: I added the "Span" to clarify that they just give sections of the deltas tables including null elements, rather than gathering the current valid ones oslt, but the type does indicate that too.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs:939]: It does essentially and that's a better name.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:1021]: Right, that's a good idea. We already did that in `GetFixedLength`, so I'll change it here too.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs:2201]: That's a better way to put it, yes.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs
Comment (olsaarik) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/DfaMatchingState.cs:32]: Likely, but I don't for sure know yet. I'll measure against current main after I get this merged (to unblock Margus).

**Comment by AndyAyersMS** [comment]
Possible regressions:
* https://github.com/dotnet/perf-autofiling-issues/issues/5761

**Comment by mrsharm** [comment]
@olsaarik - we found the following regressions that seemed to line up with this PR and as per the issues referenced by @AndyAyersMS. We detected this from our analysis while creating the perf report for August. Would you consider these regressions as "by design"? 

We did notice an improvement, however, we are still not at the same level as before this change: 

image

CC: @dakersnar 

<details>

System.Text.RegularExpressions.Tests.Perf_Regex_Industry_BoostDocs_Simple.IsMatch(Id: 11, Options: NonBacktracking)

| Result |Ratio | Alloc Delta | Operating System      | Bit   | Processor Name                                  | 
| ------ |-----:| -----------:| --------------------- | ----- | ----------------------------------------------- | 
| Slower | 0.88 |          +0 | Windows 11            | Arm64 | Microsoft SQ1 3.0 GHz                           | 
| Same   | 0.92 |          +0 | Windows 11            | Arm64 | Microsoft SQ1 3.0 GHz                           | 
| Slower | 0.81 |          +0 | macOS Monterey 12.3   | Arm64 | Apple M1 Max                                    | 
| Same   | 0.91 |          +0 | Windows 10            | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz               | 
| Slower | 0.87 |          +0 | Windows 10            | X64   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        | 
| Slower | 0.86 |          +0 | Windows 10            | X64   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        | 
| Slower | 0.88 |          +0 | Windows 10            | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R)   | 
| Slower | 0.86 |          +0 | Windows 10            | X64   | Intel Core i9-10900K CPU 3.70GHz                | 
| Slower | 0.84 |          +0 | Windows 11            | X64   | AMD Ryzen Threadripper PRO 3945WX 12-Cores      | 
| Slower | 0.85 |          +0 | Windows 11            | X64   | AMD Ryzen 9 3950X                               | 
| Same   | 0.93 |          +0 | Windows 11            | X64   | AMD Ryzen 9 5900X                               | 
| Same   | 0.94 |          +0 | Windows 11            | X64   | AMD Ryzen 9 5950X                               | 
| Same   | 0.89 |          +0 | Windows 11            | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)    | 
| Slower | 0.87 |          +0 | Windows 11            | X64   | Intel Core i9-10900K CPU 3.70GHz                | 
| Slower | 0.79 |          +0 | Windows 11            | X64   | 11th Gen Intel Core i9-11900H 2.50GHz           | 
| Slower | 0.74 |          +0 | ubuntu 18.04          | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz               | 
| Slower | 0.88 |          +0 | ubuntu 18.04          | X64   | Intel Core i7-2720QM CPU 2.20GHz (Sandy Bridge) | 
| Slower | 0.62 |          +0 | ubuntu 18.04          | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)    | 
| Slower | 0.62 |          +0 | ubuntu 20.04          | X64   | AMD Ryzen 9 5900X                               | 
| Slower | 0.78 |          +0 | ubuntu 20.04          | X64   | Intel Core i9-10900K CPU 3.70GHz                | 
| Slower | 0.85 |          +0 | Windows 10            | X86   | Intel Xeon CPU E5-1650 v4 3.60GHz               | 
| Slower | 0.78 |          +0 | Windows 10            | X86   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        | 
| Slower | 0.81 |          +0 | Windows 11            | X87   | AMD Ryzen Threadripper PRO 3945WX 12-Cores      | 
| Slower | 0.78 |          +0 | macOS Big Sur 11.6.8  | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)       | 
| Slower | 0.81 |          +0 | macOS Monterey 12.3.1 | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)     | 
| Slower | 0.76 |          +0 | macOS Monterey 12.4   | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)       | 

System.Text.RegularExpressions.Tests.Perf_Regex_Industry_BoostDocs_Simple.IsMatch(Id: 0, Options: NonBacktracking)

| Result |Ratio | Alloc Delta | Operating System      | Bit   | Processor Name                                  |
| ------ |-----:| -----------:| --------------------- | ----- | ----------------------------------------------- |
| Same   | 0.93 |          +0 | Windows 11            | Arm64 | Microsoft SQ1 3.0 GHz                           |
| Same   | 0.93 |          +0 | Windows 11            | Arm64 | Microsoft SQ1 3.0 GHz                           |
| Same   | 0.94 |          +0 | macOS Monterey 12.3   | Arm64 | Apple M1 Max                                    |
| Same   | 0.92 |          +0 | Windows 10            | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz               |
| Slower | 0.90 |          +0 | Windows 10            | X64   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        |
| Same   | 0.96 |          +0 | Windows 10            | X64   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        |
| Same   | 0.94 |          +0 | Windows 10            | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R)   |
| Same   | 0.94 |          +0 | Windows 10            | X64   | Intel Core i9-10900K CPU 3.70GHz                |
| Same   | 0.91 |          +0 | Windows 11            | X64   | AMD Ryzen Threadripper PRO 3945WX 12-Cores      |
| Slower | 0.88 |          +0 | Windows 11            | X64   | AMD Ryzen 9 3950X                               |
| Same   | 0.98 |          +0 | Windows 11            | X64   | AMD Ryzen 9 5900X                               |
| Same   | 1.02 |          +0 | Windows 11            | X64   | AMD Ryzen 9 5950X                               |
| Same   | 0.96 |          +0 | Windows 11            | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)    |
| Same   | 0.95 |          +0 | Windows 11            | X64   | Intel Core i9-10900K CPU 3.70GHz                |
| Same   | 0.91 |          +0 | Windows 11            | X64   | 11th Gen Intel Core i9-11900H 2.50GHz           |
| Same   | 0.91 |          +0 | ubuntu 18.04          | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz               |
| Same   | 0.97 |          +0 | ubuntu 18.04          | X64   | Intel Core i7-2720QM CPU 2.20GHz (Sandy Bridge) |
| Slower | 0.76 |          +0 | ubuntu 18.04          | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)    |
| Slower | 0.82 |          +0 | ubuntu 20.04          | X64   | AMD Ryzen 9 5900X                               |
| Same   | 0.90 |          +0 | ubuntu 20.04          | X64   | Intel Core i9-10900K CPU 3.70GHz                |
| Same   | 0.89 |          +0 | Windows 10            | X86   | Intel Xeon CPU E5-1650 v4 3.60GHz               |
| Same   | 0.91 |          +0 | Windows 10            | X86   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        |
| Slower | 0.85 |          +0 | Windows 11            | X86   | AMD Ryzen Threadripper PRO 3945WX 12-Cores      |
| Slower | 0.85 |          +0 | macOS Big Sur 11.6.8  | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)       |
| Slower | 0.89 |          +0 | macOS Monterey 12.3.1 | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)     |
| Same   | 0.91 |          +0 | macOS Monterey 12.4   | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)       |

System.Text.RegularExpressions.Tests.Perf_Regex_Common.CtorInvoke(Options: Compiled)

| Result |Ratio | Alloc Delta | Operating System      | Bit   | Processor Name                                  |
| ------ |-----:| -----------:| --------------------- | ----- | ----------------------------------------------- |
| Slower | 0.84 |        +551 | Windows 11            | Arm64 | Microsoft SQ1 3.0 GHz                           |
| Slower | 0.80 |        +539 | Windows 11            | Arm64 | Microsoft SQ1 3.0 GHz                           |
| Same   | 1.09 |        +550 | macOS Monterey 12.3   | Arm64 | Apple M1 Max                                    |
| Slower | 0.87 |        +542 | Windows 10            | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz               |
| Slower | 0.80 |        +552 | Windows 10            | X64   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        |
| Slower | 0.64 |        +553 | Windows 10            | X64   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        |
| Slower | 0.83 |        +562 | Windows 10            | X64   | Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R)   |
| Slower | 0.88 |        +534 | Windows 10            | X64   | Intel Core i9-10900K CPU 3.70GHz                |
| Slower | 0.89 |        +559 | Windows 11            | X64   | AMD Ryzen Threadripper PRO 3945WX 12-Cores      |
| Same   | 0.89 |        +561 | Windows 11            | X64   | AMD Ryzen 9 3950X                               |
| Same   | 0.93 |        +520 | Windows 11            | X64   | AMD Ryzen 9 5900X                               |
| Same   | 0.93 |        +519 | Windows 11            | X64   | AMD Ryzen 9 5950X                               |
| Slower | 0.88 |        +543 | Windows 11            | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)    |
| Slower | 0.86 |        +539 | Windows 11            | X64   | Intel Core i9-10900K CPU 3.70GHz                |
| Slower | 0.86 |        +558 | Windows 11            | X64   | 11th Gen Intel Core i9-11900H 2.50GHz           |
| Slower | 0.88 |        +543 | ubuntu 18.04          | X64   | Intel Xeon CPU E5-1650 v4 3.60GHz               |
| Same   | 1.02 |        +548 | ubuntu 18.04          | X64   | Intel Core i7-2720QM CPU 2.20GHz (Sandy Bridge) |
| Slower | 0.77 |        +554 | ubuntu 18.04          | X64   | Intel Core i7-8700 CPU 3.20GHz (Coffee Lake)    |
| Same   | 0.90 |        +557 | ubuntu 20.04          | X64   | AMD Ryzen 9 5900X                               |
| Slower | 0.82 |        +539 | ubuntu 20.04          | X64   | Intel Core i9-10900K CPU 3.70GHz                |
| Slower | 0.86 |        +544 | Windows 10            | X86   | Intel Xeon CPU E5-1650 v4 3.60GHz               |
| Slower | 0.77 |        +555 | Windows 10            | X86   | Intel Core i7-6700 CPU 3.40GHz (Skylake)        |
| Slower | 0.89 |        +545 | Windows 11            | X86   | AMD Ryzen Threadripper PRO 3945WX 12-Cores      |
| Slower | 0.76 |        +532 | macOS Big Sur 11.6.8  | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)       |
| Slower | 0.79 |        +548 | macOS Monterey 12.3.1 | X64   | Intel Core i7-5557U CPU 3.10GHz (Broadwell)     |
| Slower | 0.79 |        +543 | macOS Monterey 12.4   | X64   | Intel Core i5-4278U CPU 2.60GHz (Haswell)       |

</details>

**Comment by stephentoub** [comment]
(While this could have affected the two NonBacktracking tests, this would not have affected the Compiled test.)

**Pull request merged**
